components/dashboard/src/Login.tsx

```diff
@@ -138,20 +190,10 @@ export function Login() {
                         <div className="mx-auto pb-8">
                             <img src={gitpodIcon} className="h-16 mx-auto" alt="Gitpod's logo" />
                         </div>
-                        <div className="mx-auto text-center pb-8 space-y-2">
-                            <h1 className="text-3xl">Log in{showWelcome ? '' : ' to Gitpod'}</h1>
-                            <h2 className="uppercase text-sm text-gray-400">ALWAYS READY-TO-CODE</h2>
-                        </div>
-                        <div className="flex flex-col space-y-3 items-center">
-                            {authProviders.map(ap => {
-                                return (
-                                    <button key={"button" + ap.host} className="btn-login flex-none w-56 h-10 p-0 inline-flex" onClick={() => openLogin(ap.host)}>
-                                        {iconForAuthProvider(ap.authProviderType)}
-                                        <span className="pt-2 pb-2 mr-3 text-sm my-auto font-medium truncate overflow-ellipsis">Continue with {simplifyProviderName(ap.host)}</span>
-                                    </button>
-                                );
-                            })}
-                        </div>
+
+                        {renderWelcomeText()}
+
+                        {renderAuthProviders()}
```

Comment:
**Thought**: This is very much personal preference, but I personally prefer keeping all HTML markup together, because I think it helps quickly understanding how the page is built, and facilitates making structural changes in the future. I find that splitting the `render()` function into smaller `renderXYZ()` functions breaks apart the markup too much, and makes it harder to rework it later.

E.g. compare these pseudo examples:

```jsx
    return <>
        <Header title="my page" />
        <p>Manage your elements</p>
        {elements.length > 0
            ? <>
                <h2>My Elements</h2>
                <List elements={elements} />
            </>
            : <>
                <h2>No Elements Yet!</h2>
                <Button onClick={createElement()}>New Element</Button>
            </>}
    </>;
```

with:

```jsx
    const renderSubtitle = () => {
        return <p>Manage your elements</p>;
    }

    const renderEmptyState = () => {
        return <>
            <h2>No Elements Yet!</h2>
            <Button onClick={createElement()}>New Element</Button>
        </>;
    }

    const renderElementsList = (elements) => {
        return <>
            <h2>My Elements</h2>
            <List elements={elements} />
        </>;
    }

    const renderElementsPage = (elements) => {
        if (elements.length === 0) {
            return renderElementsList(elements);
        }
        return renderEmptyState();
    }

    return <>
        <Header title="my page" />
        {renderSubtitle()}
        {renderElementsPage()}
    </>;
```

Okay, maybe my example is a little bit contrived, but the second example is so hard to understand and navigate 🙈 you need to jump back and forth across the code, and the HTML markup is split all over the place in random order.

I'd find the first example much easier to rework (e.g. if you need to restyle some part, or move the markup under another `<div>` layer, or restructure the page a bit by making more things conditional, etc).

But again, this is totally personal, so please feel free to ignore if you do find sub-render functions better! 😁

Reply:
Good point, thank you for sharing your view on this. I know others in the team prefer split render functions as well (hi @AlexTugarev 👋) so I might be in the minority here. I'll make more efforts to adapt. 😁